### PR TITLE
Update backend to support accepting headers at resource level

### PR DESCRIPTION
**Pull request by omerdemirok**
Signed-off-by: Omer Demirok <odemirok@chef.io>

### Description

- Update backend to support accepting HTTP request headers at the resource level
- Add a helper module for hash keys modification recursively

### Issues Resolved

Unblocks #324 

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: libraries/backend/azure_connection.rb [hunk 1]
```diff
@@ -157,7 +162,10 @@ def fail_api_query(resp, message = nil)
     message ||= "Unsuccessful HTTP request to Azure REST API.\n"
     message += "HTTP #{resp.status}.\n"
     body = resp.body
-    unless body.empty?
+    if body.empty?
```
Comment (rmoles) [inline libraries/backend/azure_connection.rb:165]: Is this change intentional? I would have imagined the message body can have content when successful also.

Comment (omerdemirok) [reply libraries/backend/azure_connection.rb:165]: This is in the `fail_api_query` method which is called only when the HTTP status is not 200.
It is just a refactoring to prevent increasing the conditional statement depth. The functionality is the same.

Comment (rmoles) [reply libraries/backend/azure_connection.rb:165]: cool no worries

**Review by rmoles — APPROVED** [review]
Looks good to me
